src/librustc_driver/args/mod.rs

```diff
@@ -0,0 +1,207 @@
+                Some(filearg) => return Some(filearg),
+                None => {
+                    let _ = self.stack.pop();
+                }
+            }
+        }
+
+        None
+    }
+}
+
+impl Iterator for ArgsStack {
+    type Item = OsString;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            let arg = self.stack_next().or_else(|| self.base.next().map(FileArg::from));
+            match arg {
+                None => return None,
+                Some(FileArg::Arg(arg)) => return Some(arg),
+                Some(FileArg::Path(path)) => {
+                    let file = if self.stack.len() >= MAX_RECURSION {
+                        eprintln!(
+                            "Warning: Max recursion depth {} reached for file {}",
+                            MAX_RECURSION,
+                            path.display()
+                        );
+                        None
+                    } else {
+                        match fs::read(&path) {
+                            Ok(file) => Some(file),
+                            Err(err) => {
+                                eprintln!(
+                                    "Warning: Failed to open {}: {} (cwd: {})",
+                                    path.display(),
+                                    err,
+                                    env::current_dir().unwrap().display(),
+                                );
+                                None
+                            }
+                        }
+                    };
+                    if let Some(file) = file {
+                        self.stack.push(Box::new(file_args(file)))
+                    } else {
+                        // Failed to open - try to reconstruct original arg
+                        let arg = format!("@{}", path.display());
+                        return Some(OsString::from(arg));
+                    }
+                }
+            }
+        }
+    }
+}
+
+#[cfg(any(unix, target_os = "redox"))]
+fn os_string_from_bytes(s: &[u8]) -> OsString {
+    use std::os::unix::ffi::OsStrExt;
+    std::ffi::OsStr::from_bytes(s).to_os_string()
+}
+#[cfg(not(any(unix, target_os = "redox")))]
+fn os_string_from_bytes(s: &[u8]) -> OsString {
+    OsString::from(&*String::from_utf8_lossy(s))
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+enum FileArg {
+    Arg(OsString),
+    Path(PathBuf),
+}
+
+impl From<Vec<u8>> for FileArg {
+    fn from(v: Vec<u8>) -> Self {
+        if v.starts_with(b"@") {
+            FileArg::Path(PathBuf::from(os_string_from_bytes(&v[1..])))
+        } else {
+            FileArg::Arg(OsString::from(os_string_from_bytes(&v)))
+        }
+    }
+}
+
+impl From<OsString> for FileArg {
+    fn from(v: OsString) -> Self {
+        FileArg::from(v.to_string_lossy().into_owned().into_bytes())
+    }
+}
+
+fn file_args(file: Vec<u8>) -> impl Iterator<Item = FileArg> {
+    #[derive(Copy, Clone, Debug, Eq, PartialEq)]
+    enum Quote {
```

Comment:
Personally I feel like having a whole quoting system is overkill. I understand that other tools are doing this, but is there really that much shared tooling we're trying to leverage here? It seem reasonable to me to just say that we accept on argument per line, meaning we just split on `\n` characters (or even `\r\n` as well). You brought up concerns about line endings but I feel like those are much more easily handled than adding a whole quoting parser here simply for command line arguments.

Reply:
Newline delimiting would prevent an option from ever having an internal newline. I guess looking at rustc's common options that doesn't seem to be a big limitation. Will that constraint also be always OK for Cargo? Would it ever make sense to have `--cfg 'foo="bar<literal-nl>blat"'`?

In general there's a question of whether the line-breaks are separators or terminators, and whether you can represent empty options. If you can't have empty options then it doesn't matter too much, but otherwise they'd be represented as a blank line. Then the question of whether its a separator or a terminator affects whether a trailing `\n` on the end of the file is an empty option (separator) or not (terminator). And if its a terminator, if you leave the trailing `\n` off is that OK (I guess).

And regarding `\n` vs `\r\n`, I guess it could implement a state machine where each of `\n`, `\r\n` and lone `\r` are all treated as a single line end.

But overall, it still feels like this is all much more awkward for users to use. Correctly generating a line-delimited file requires the generator to get some fairly subtle details correct, and getting them wrong could be awkward to diagnose and understand. The tradeoff would be a reduction of complexity in this code, but it is pretty localized complexity, and I think having it makes using this feature much more straightforward.

And of course this is all influenced by the fact that we do already have quite a bit of tooling which knows how to deal with these kinds of files (Buck has common code for it, for example), and if the Rust ecosystem chooses something else then it would need to have parallel implementations.

Reply:
If a newline doesn't work, then perhaps we could use the nul byte? You mentioned it is difficult to generate from a shell script, but does that really happen that often?

I'm sort of under the impression that all users of this functionality are build systems which are generating these files anyway. In that sense using something simple like a nul byte instead of a whole quoting system (which then involves figuring out what to quote and all that) would be a lot easier and still solve all our problems.

AFAIK there's basically no real use case for passing a `\n` in an argument, and if there really is we could perhaps invent syntax for it later like `@+file` which means "literally include the contents of `file` as an argument" or something weird like that.

I'm pretty hesitant to blindly follow what binutils/gnu stuff does without a good reason to do so.

Reply:
Yes, generating from a shell script would be a primary use case.

I think the options forward are:
1. as-is, with bespoke quoting
2. use the shlex crate, so that its at least easy to produce/parse elsewhere in the ecosystem
3. `\n` separated if we're OK with not allowing `\n` in options (and probably not allowing empty-string options)
4. `\0` which has precedent (xargs -0) but is otherwise pretty awkward for the use cases I see for this.